webapp/src/Service/DOMJudgeService.php

```diff
@@ -993,13 +998,50 @@ public function unblockJudgeTasksForProblem(int $probId): void
         }
     }
 
-    public function maybeCreateJudgeTasks(Judging $judging, int $priority = JudgeTask::PRIORITY_DEFAULT): void
+    public function unblockJudgeTasksForSubmission(string $submissionId): void
+    {
+        // These are all the judgings that don't have associated judgetasks yet. Check whether we unblocked them.
+        $judgings = $this->helperUnblockJudgeTasks()
+            ->join(Submission::class, 's', Join::WITH, 'j.submission = s.submitid')
+            ->andWhere('j.submission = :submissionid')
+            ->setParameter('submissionid', $submissionId)
+            ->getQuery()
+            ->getResult();
+        foreach ($judgings as $judging) {
+            $this->maybeCreateJudgeTasks($judging, JudgeTask::PRIORITY_DEFAULT, True);
+        }
+    }
+
+    public function unblockJudgeTasks(): void
+    {
+        // These are all the judgings that don't have associated judgetasks yet. Check whether we unblocked them.
+        $judgings = $this->helperUnblockJudgeTasks()
+            ->getQuery()
+            ->getResult();
+        foreach ($judgings as $judging) {
+            $this->maybeCreateJudgeTasks($judging);
+        }
+    }
+
+    public function maybeCreateJudgeTasks(Judging $judging, int $priority = JudgeTask::PRIORITY_DEFAULT, bool $manualRequest = False): void
     {
         $submission = $judging->getSubmission();
         $problem    = $submission->getContestProblem();
         $language   = $submission->getLanguage();
 
-        if (!$problem->getAllowJudge() || !$language->getAllowJudge()) {
+        $evalOnDemand = False;
+        // We have 2 cases, the problem picks the global value or the value is set.
+        if ( ((int)$problem->getLazyEvalResults() === 0 && $this->config->get('lazy_eval_results') === static::EVAL_DEMAND)
+             || $problem->getLazyEvalResults() === 3) {
+            $evalOnDemand = True;
+        }
+        // Special case, we're shadow and someone submits on our side in that case
+        // we're not super lazy.
+        if ($this->config->get('data_source') === DOMJudgeService::DATA_SOURCE_CONFIGURATION_AND_LIVE_EXTERNAL
+            && $submission->getExternalid() === null) {
+                $evalOnDemand = False;
+        }
+        if ((!$problem->getAllowJudge() || !$language->getAllowJudge() || $evalOnDemand) && !$manualRequest) {
```

Comment:
I think we should not judge this problem if the language or the problem are disabled.

Basically I would remove the `&& $!manualRequest` and in the line before do something like:
```
// Submission was requested explicitly to be judged.
if ($manualRequest) {
  $evalOnDemand = false;
}
```

(please check my logic)

Reply:
So I'm not 100% sure who's wrong here, but I'll just try to rewrite this to a form where it is easier to understand who is correct as those negations combined with `OR` is difficult to read.

Reply:
Implemented